### PR TITLE
feat: Add otlp metrics exporter and update prom exporter

### LINKED ISSUE #1647: Feat: Add Otlp metrics exporter to enable Strato Integration
**Describe the solution you'd like**
I would like to be able to specify otlp in the command line along with an endpoint and if its insecure and have the csi driver use that exporter with the corresponding parameters. 

**Anything else you would like to add:**
This will enable the csi driver to integrate with Strato, and allow projects that use this driver to moniter the corresponding azure calls, enabling things like billing.

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Azure Key Vault provider version: (use the image tag):
- Kubernetes version: (use `kubectl version`):
- Cluster type: (e.g. AKS, aks-engine, etc):

### PR DESCRIPTION
**Pull request by AlexanderStocks**
**Reason for Change**:

This adds the otlp exporter with the environment variables required to configure strato, or any other generic grpc endpoint.
This also updates the otel library which includes breaking changes to the Prometheus exporter

**Requirements**

- [ ] squashed commits
- [ ] included documentation
- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

Fixes #1647 
**Does this change contain code from or inspired by another project?**

- [x] Yes
- [ ] No

**If "Yes," did you notify that project's maintainers and provide attribution?**
Its my code.
**Special Notes for Reviewers**:

Output from /metrics:

```
# HELP grpc_request Distribution of how long it took for the gRPC requests
# TYPE grpc_request histogram
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.1"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.2"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.3"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.4"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.5"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="1"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="1.5"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="2"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="2.5"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="3"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="5"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="10"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="15"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="30"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="+Inf"} 1
grpc_request_sum{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure"} 0.000118976
grpc_request_count{grpc_code="OK",grpc_message="",grpc_method="/grpc.health.v1.Health/Check",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.1"} 0
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.2"} 0
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.3"} 0
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.4"} 0
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.5"} 0
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="1"} 1
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="1.5"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="2"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="2.5"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="3"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="5"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="10"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="15"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="30"} 2
grpc_request_bucket{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="+Inf"} 2
grpc_request_sum{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure"} 2.2844812670000003
grpc_request_count{grpc_code="OK",grpc_message="",grpc_method="/v1alpha1.CSIDriverProvider/Mount",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure"} 2
# HELP keyvault_request Distribution of how long it took to get from keyvault
# TYPE keyvault_request histogram
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.1"} 0
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.2"} 0
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.3"} 0
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.4"} 0
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="0.5"} 0
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="1"} 1
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="1.5"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="2"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="2.5"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="3"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="5"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="10"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="15"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="30"} 2
keyvault_request_bucket{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure",le="+Inf"} 2
keyvault_request_sum{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure"} 2.282627427
keyvault_request_count{error="",object_name="bugbash23sscontrolleras",object_type="secret",os_type="linux",otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version="",provider="azure",service_name="csi-secrets-store-provider-azure"} 2
# HELP otel_scope_info Instrumentation Scope metadata
# TYPE otel_scope_info gauge
otel_scope_info{otel_scope_name="csi-secrets-store-provider-azure",otel_scope_version=""} 1
# HELP promhttp_metric_handler_errors_total Total number of internal errors encountered by the promhttp metric handler.
# TYPE promhttp_metric_handler_errors_total counter
promhttp_metric_handler_errors_total{cause="encoding"} 0
promhttp_metric_handler_errors_total{cause="gathering"} 0
# HELP target_info Target metadata
# TYPE target_info gauge
target_info{service_name="unknown_service:secrets-store-csi-driver-provider-azure",telemetry_sdk_language="go",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="1.29.0"} 1
```

### CODE REVIEW

File: pkg/metrics/otlp_exporter.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		return errors.Errorf("OTLP exporter specified but endpoint not provided")
+	}
+
+	otlpOpts := []otlpmetricgrpc.Option{
+		otlpmetricgrpc.WithEndpoint(otlpMetricsEndpointGrpc),
+		otlpmetricgrpc.WithRetry(otlpmetricgrpc.RetryConfig{
+			Enabled: false,
+		}),
+		otlpmetricgrpc.WithTemporalitySelector(func(kind metric.InstrumentKind) metricdata.Temporality {
+			return metricdata.DeltaTemporality
```
Comment (nilekhc) [inline pkg/metrics/otlp_exporter.go:51]: Is there any rationale about using this over CumulativeTemporality?

**Review by nilekhc — CHANGES_REQUESTED** [review]
Could you provide metrics output from grpc endpoint? is it same as in pr description?

File: pkg/metrics/otlp_exporter.go (hunk 1 above)
Comment (AlexanderStocks) [reply pkg/metrics/otlp_exporter.go:51]: Strato doesnt work with CumulativeTemporality.

File: pkg/metrics/otlp_exporter.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+var (
+	otlpMetricsEndpointGrpc = os.Getenv("otlpMetricsEndpointGrpc")
+	extensionResourceID     = os.Getenv("EXTENSION_RESOURCE_ID")
+	otlpInsecureGrpc        = os.Getenv("otlpInsecureGrpc")
```
[the remark this answers is not in the document]
Comment (AlexanderStocks) [reply pkg/metrics/otlp_exporter.go:35]: until last week strato only supported insecure. I am in the process of changing this to secure. Insecure is being deprecated so I will update this PR when I have validated my change works.

**Pull request closed without merging**